Firestore/core/test/firebase/firestore/local/index_free_query_engine_test.cc

```diff
@@ -0,0 +1,336 @@
+                                          Map("matches", false, "order", 1),
+                                          DocumentState::kLocalMutations);
+Document UPDATED_DOC_A = Doc("coll/a", 11, Map("matches", true, "order", 1));
+Document MATCHING_DOC_B = Doc("coll/b", 1, Map("matches", true, "order", 2));
+Document UPDATED_MATCHING_DOC_B =
+    Doc("coll/b", 11, Map("matches", true, "order", 2));
+
+SnapshotVersion LAST_LIMBO_FREE_SNAPSHOT = Version(10);
+SnapshotVersion MISSING_LAST_LIMBO_FREE_SNAPSHOT = SnapshotVersion::None();
+
+}  // namespace
+
+class TestLocalDocumentsView : public LocalDocumentsView {
+ public:
+  TestLocalDocumentsView(RemoteDocumentCache* remote_document_cache,
+                         MutationQueue* mutation_queue,
+                         IndexManager* index_manager)
+      : LocalDocumentsView(
+            remote_document_cache, mutation_queue, index_manager) {
+  }
+
+  model::DocumentMap GetDocumentsMatchingQuery(
+      const core::Query& query,
+      const model::SnapshotVersion& since_read_time) override {
+    EXPECT_TRUE(expect_index_free_execution.has_value());
+    EXPECT_EQ(*expect_index_free_execution,
+              since_read_time != SnapshotVersion::None());
+    return LocalDocumentsView::GetDocumentsMatchingQuery(query,
+                                                         since_read_time);
+  }
+
+  void ExpectIndexFreeExecution(bool index_free) {
+    expect_index_free_execution = index_free;
+  }
+
+ private:
+  absl::optional<bool> expect_index_free_execution;
+};
+
+class IndexFreeQueryEngineTest : public ::testing::Test {
+ public:
+  IndexFreeQueryEngineTest()
+      : persistence_(MemoryPersistence::WithEagerGarbageCollector()),
+        remote_document_cache_(persistence_->remote_document_cache()),
+        query_cache_(persistence_->query_cache()),
+        index_manager_(new MemoryIndexManager()),
+        local_documents_view_(
+            remote_document_cache_,
+            persistence_->GetMutationQueueForUser(User::Unauthenticated()),
+            index_manager_.get()) {
+    query_engine_.SetLocalDocumentsView(&local_documents_view_);
+  }
+
+  /** Adds the provided documents to the query target mapping. */
+  void PersistQueryMapping(const std::vector<DocumentKey>& keys) {
+    persistence_->Run("PersistQueryMapping", [this, &keys]() {
+      DocumentKeySet remote_keys;
+      for (const DocumentKey& key : keys) {
+        remote_keys = remote_keys.insert(key);
+      }
+      query_cache_->AddMatchingKeys(remote_keys, TEST_TARGET_ID);
+    });
+  }
+
+  /** Adds the provided documents to the remote document cache. */
+  void AddDocuments(const std::vector<Document>& docs) {
+    persistence_->Run("AddDocuments", [this, &docs]() {
+      for (const Document& doc : docs) {
+        remote_document_cache_->Add(doc, doc.version());
+      }
+    });
+  }
+
+  DocumentSet ExpectIndexFreeQuery(const std::function<DocumentSet(void)>& f) {
```

Comment:
Optional: you don't necessarily have to use `std::function` here, you could make it a template function:
```
template <typename F>
DocumentSet ExpectIndexFreeQuery(const F& f) {
```
Otherwise, please include `<functional>`.

Update: I think this function could be rewritten to not use a callback at all:
```cpp
DocumentSet ExpectIndexFreeQuery(const Query& query, SnapshotVersion version) {
  local_documents_view_.ExpectIndexFreeExecution(true);
  return RunQuery(query, version);
}
```

Same applies to `ExpectFullCollectionQuery`.

Reply:
I kind of prefer `std::function` over the template type. It makes it immediately obvious to me what this method excepts, whereas `const F& f` only reveals its type via the compiler.

As for not using a callback - we settled on this approach in the other ports to make the tests more explicit. I'd prefer to keep it as is to match the other platforms.

Reply:
SGTM. I suppose it would be more readable if the template argument was named something like `Function` or `Callback`, but this isn't something I feel strongly about.